tools/make_metadata.cpp

```diff
@@ -148,84 +123,168 @@ std::vector<iceberg::types::NestedField> GetFieldsFromParquet(
-ABSL_FLAG(std::string, meta_uuid, "", "UUID for metadata.json file");
-ABSL_FLAG(bool, force_not_null, false, "make columns NOT NULL");
-ABSL_FLAG(bool, byte_array_is_string, false, "intarpret BYTE_ARRAY type as String (Binary if not set)");
+ABSL_FLAG(std::string, parquet_path, "", "path to parquet file (URI)");
+
+ABSL_FLAG(std::string, table_base_location, "", "location where data and metadata files are stored (URI)");
+ABSL_FLAG(std::string, root_snapshot_path, "", "path for .metadata.json file (URI)");
+ABSL_FLAG(std::string, table_uuid, "", "");
+
+ABSL_FLAG(std::string, filesystem, "file", "filesystem to use (file or s3)");
+ABSL_FLAG(std::string, s3_access_key_id, "", "s3_access_key_id");
+ABSL_FLAG(std::string, s3_secret_access_key, "", "s3_secret_access_key");
+ABSL_FLAG(std::string, s3_endpoint, "", "s3_endpoint");
+ABSL_FLAG(std::string, s3_scheme, "http", "s3_scheme");
+ABSL_FLAG(std::string, s3_region, "", "s3_region");
 
 int main(int argc, char** argv) {
-  try {
-    static const std::string meta_ext = ".metadata.json";
+  std::optional<S3FinalizerGuard> s3_guard;
 
+  try {
     absl::ParseCommandLine(argc, argv);
 
-    const std::string src_data = absl::GetFlag(FLAGS_src_data);
-    const std::string dst = absl::GetFlag(FLAGS_dst);
-    const std::string data_dir = absl::GetFlag(FLAGS_data_dir);
-    const std::string meta_dir = absl::GetFlag(FLAGS_meta_dir);
-    std::string meta_uuid = absl::GetFlag(FLAGS_meta_uuid);
-    const bool force_not_null = absl::GetFlag(FLAGS_force_not_null);
-    const bool byte_array_is_string = absl::GetFlag(FLAGS_byte_array_is_string);
-
-    if (src_data.empty()) {
-      throw std::runtime_error("No src specified");
-    }
-    if (dst.empty()) {
-      throw std::runtime_error("No dst directory specified");
-    }
-    if (meta_dir.empty() || std::filesystem::exists(meta_dir)) {
-      throw std::runtime_error("Wrong meta_dir (empty or exists): " + meta_dir);
-    }
-    if (meta_uuid.empty()) {
-      meta_uuid = iceberg::UuidGenerator().CreateRandom().ToString();
+    const std::string filesystem_type = absl::GetFlag(FLAGS_filesystem);
+
+    std::shared_ptr<arrow::fs::FileSystem> fs;
+    if (filesystem_type == "file") {
+      fs = std::make_shared<arrow::fs::LocalFileSystem>();
+    } else if (filesystem_type == "s3") {
+      s3_guard.emplace();
+      const std::string access_key = absl::GetFlag(FLAGS_s3_access_key_id);
+      const std::string secret_key = absl::GetFlag(FLAGS_s3_secret_access_key);
+      auto s3options = arrow::fs::S3Options::FromAccessKey(access_key, secret_key);
+      s3options.endpoint_override = absl::GetFlag(FLAGS_s3_endpoint);
+      s3options.scheme = absl::GetFlag(FLAGS_s3_scheme);
+      s3options.region = absl::GetFlag(FLAGS_s3_region);
+      auto maybe_fs = arrow::fs::S3FileSystem::Make(s3options);
+      if (!maybe_fs.ok()) {
+        std::cerr << maybe_fs.status() << std::endl;
+        return 1;
```

Comment:
It's better to throw here. It leads to fewer changes If some day we replace std::cerr with logger or move the code out of main/binary to library